pkg/pillar/agentbase/agent.go

```diff
@@ -15,54 +15,114 @@ import (
 type AgentBase interface {
 	AgentBaseContext() *Context
-	AddAgentSpecificCLIFlags()
-	ProcessAgentSpecificCLIFlags()
+	AddAgentSpecificCLIFlags(*flag.FlagSet)
+	ProcessAgentSpecificCLIFlags(*flag.FlagSet)
 }
 
 // CliParams - stores all the common cli params
 type CliParams struct {
 	DebugOverride bool
 }
 
+// ContextOpt function may be used to modify options
+type ContextOpt func(*Context)
+
+// WithPidFile creates pid file
+func WithPidFile() ContextOpt {
+	return func(context *Context) {
+		context.needPidFile = true
+	}
+}
+
+// WithArguments defines arguments
+func WithArguments(arguments []string) ContextOpt {
+	return func(context *Context) {
+		context.arguments = arguments
+	}
+}
+
+// WithWatchdog defines options required to initialize watchdog
+func WithWatchdog(pubSub *pubsub.PubSub, warningTime, errorTime time.Duration) ContextOpt {
+	return func(context *Context) {
+		context.pubSub = pubSub
+		context.warningTime = warningTime
+		context.errorTime = errorTime
+		context.needWatchdog = true
+	}
+}
+
 // Context - a struct that represents a general agent context
 type Context struct {
-	PubSub       *pubsub.PubSub
-	Logger       *logrus.Logger
-	Log          *base.LogObject
-	CLIParams    CliParams
-	ErrorTime    time.Duration
-	WarningTime  time.Duration
-	AgentName    string
-	NeedWatchdog bool
+	pubSub       *pubsub.PubSub
+	logger       *logrus.Logger
+	log          *base.LogObject
+	cliParams    CliParams
+	errorTime    time.Duration
+	warningTime  time.Duration
+	agentName    string
+	needWatchdog bool
+	needPidFile  bool
+	arguments    []string
+}
+
+// CLIParams returns CliParams
+func (ctx *Context) CLIParams() CliParams {
+	return ctx.cliParams
+}
+
+// Logger returns Logger
+func (ctx *Context) Logger() *logrus.Logger {
+	return ctx.logger
+}
+
+// InitAgentBaseContext initialize Context with defined options
+func InitAgentBaseContext(logger *logrus.Logger, log *base.LogObject, agentName string, opts ...ContextOpt) *Context {
+	ctx := &Context{
+		logger:    logger,
+		log:       log,
+		agentName: agentName,
+	}
+	for _, opt := range opts {
+		opt(ctx)
+	}
+	return ctx
 }
 
 // processCLIFlags - Add flags common to all agents
-func processCLIFlags(agentBase AgentBase) {
-	debugPtr := flag.Bool("d", false, "Debug flag")
-	agentBase.AddAgentSpecificCLIFlags()
-	flag.Parse()
+func processCLIFlags(agentBase AgentBase) error {
 	ctx := agentBase.AgentBaseContext()
-	ctx.CLIParams.DebugOverride = *debugPtr
-	agentBase.ProcessAgentSpecificCLIFlags()
+	flagSet := flag.NewFlagSet(ctx.agentName, flag.ExitOnError)
+	debugPtr := flagSet.Bool("d", false, "Debug flag")
+	agentBase.AddAgentSpecificCLIFlags(flagSet)
+	if err := flagSet.Parse(ctx.arguments); err != nil {
+		return err
+	}
+	ctx.cliParams.DebugOverride = *debugPtr
+	agentBase.ProcessAgentSpecificCLIFlags(flagSet)
+	return nil
 }
 
 // Run - a general run function that will handle all of the common code for agents
 func Run(agentSpecificContext AgentBase) {
```

Comment:
I would suggest to rename this function from `Run` to `Init` or something like that. When reading `agentbase.Run(clientCtx)`, called inside each microservice, it gives me a wrong impression that something is actually started and running inside, but there is really just some common pre-run initialization.

Reply:
Make sense, modified to Init, it is mostly about initialization.